Investigate HIP 5.2.0 error with Kokkos 4.0

HIP 5.2.0 error:
```
fatal error: error in backend: Cannot select: 0x559739e39148: i32,ch = 
AtomicLoadAdd<(load store syncscope("agent-one-as") monotonic (s32) on %ir.161, 
addrspace 5)> 0x5597327b0b68, 0x55972efc81a8, Constant:i32<1>,
/opt/rocm-5.2.0/include/hip/amd_detail/amd_hip_atomic.h:141:10 @[
/opt/kokkos/include/desul/atomics/Fetch_Op_HIP.hpp:48:159 @[
/opt/kokkos/include/desul/atomics/Generic.hpp:327:3 @[
/opt/kokkos/include/Kokkos_Atomics_Desul_Wrapper.hpp:121:44 @[
test/tstQueryTreeRay.cpp:146:28 @[
src/details/ArborX_DetailsTreeTraversal.hpp:438:7 @[
/opt/kokkos/include/HIP/Kokkos_HIP_Parallel_Range.hpp:58:5 @[
/opt/kokkos/include/HIP/Kokkos_HIP_Parallel_Range.hpp:73:22 @[
/opt/kokkos/include/HIP/Kokkos_HIP_KernelLaunch.hpp:80:3
] ] ] ] ] ] ] ]
  0x55972efc81a8: i32 = add 0x55973360e200, Constant:i32<104>
    0x55973360e200: i32 = add FrameIndex:i32<0>, 0x55972c2dc648
      0x5597348aba68: i32 = FrameIndex<0>
      0x55972c2dc648: i32 = shl 0x559735c47df0, Constant:i32<2>
        0x559735c47df0: i32 = truncate 0x5597285b60a0
          0x5597285b60a0: i64 = AssertSext 0x559735c47d88, ValueType:ch:i32
            0x559735c47d88: i64,ch = CopyFromReg 0x5597327b0b68, Register:i64 %11
              0x55972efc8b00: i64 = Register %11
        0x55972c2dc8b8: i32 = Constant<2>
    0x55973360de58: i32 = Constant<104>
  0x55972b981710: i32 = Constant<1>
In function: _ZN6Kokkos4ImplL32hip_parallel_launch_local_memoryINS0_11ParallelForIN6ArborX7Details13TreeTraversalINS3_28BasicBoundingVolumeHierarchyINS_8HIPSpaceENS3_3BoxEvEENS4_12PermutedDataI28BoxesIntersectedByRayOrderedINS_6DeviceINS_3HIPES7_EEENS_4ViewIPjJSE_EEELb0EEE19InsertIntersectionsISE_ENS3_12Experimental26OrderedSpatialPredicateTagEEENS_11RangePolicyIJSD_NSO_11OneLeafTreeEEEESD_EELj1024ELj1EEEvT_
clang-14: note: diagnostic msg: Error generating preprocessed source(s).
```
Corresponding code: https://github.com/arborx/ArborX/blob/0136f9683051a298b5cbeeb2a940133f279042c9/test/tstQueryTreeRay.cpp#L146

_Originally posted by @aprokop in https://github.com/arborx/ArborX/issues/834#issuecomment-1454172350_

Using two different `int` instead of `count` fixes the error but now I have another problem
```
arborx/test/tstQueryTreeRay.cpp:220:47: error: no viable conversion from 'Kokkos::View<int *[2], Device<HIP, HIPSpace>>' to 'int'   
      InsertIntersections<DeviceType>{{0, 0}, device_ordered_intersections});  
                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~  
```

--- a/test/tstQueryTreeRay.cpp
+++ b/test/tstQueryTreeRay.cpp
@@ -135,7 +135,13 @@ BOOST_AUTO_TEST_SUITE_END()
 template <typename DeviceType>
 struct InsertIntersections
 {
+  // With ROCm version 5.2, we need to use a Kokkos::View to avoid a compiler
+  // bug. See https://github.com/arborx/ArborX/issues/835
+#if (HIP_VERSION_MAJOR == 5) && (HIP_VERSION_MINOR == 2)
+  Kokkos::View<int[2], DeviceType> count;
+#else
   mutable int count[2];
+#endif
   Kokkos::View<int *[2], DeviceType> _ordered_intersections;
 
   template <typename Predicate>
@@ -203,9 +209,17 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(test_ray_box_intersection_new, DeviceType,
 
   BoxesIntersectedByRayOrdered<DeviceType> predicates{device_rays};
 
+#if (HIP_VERSION_MAJOR == 5) && (HIP_VERSION_MINOR == 2)
+  Kokkos::View<int[2], DeviceType> count("count");
+  tree.query(
+      exec_space, predicates,
+      InsertIntersections<DeviceType>{count, device_ordered_intersections});
+#else
   tree.query(
       exec_space, predicates,
       InsertIntersections<DeviceType>{{0, 0}, device_ordered_intersections});
+#endif
+
   auto const host_ordered_intersections = Kokkos::create_mirror_view_and_copy(
       Kokkos::HostSpace{}, device_ordered_intersections);
   for (unsigned int i = 0; i < n; ++i)

